src/main/scala/hydrozoa/l1/rulebased/onchain/scalar/Scalar.scala

```diff
@@ -0,0 +1,117 @@
+package hydrozoa.l1.rulebased.onchain.scalar
+
+import scalus.builtin.Builtins.{byteStringToInteger, integerToByteString}
+import scalus.builtin.ByteString
+import scalus.prelude.Option
+import scalus.prelude.Option.{None, Some}
+import scalus.{Compile, Ignore}
+
+import java.math.BigInteger
+import scala.annotation.tailrec
+
+case class Scalar(private val unScalar: BigInt)
+
+@Compile
+object Scalar:
+
+    // The prime number defining the scalar field of the BLS12-381 curve.
+    val fieldPrime: BigInt = BigInt(
+      "52435875175126190479447740508185965837690552500527637822603658699938581184513"
+    )
+
+    // Constants
+    def zero: Scalar = new Scalar(BigInt(0))
+
+    def one: Scalar = new Scalar(BigInt(1))
+
+    // Construction
+    def apply(n: BigInt): Option[Scalar] =
+        if n >= 0 && n < fieldPrime then Some(new Scalar(n))
+        else None
+
+    // TODO: Why doesn't it work?
+    @Ignore
+    def apply(s: String): Option[Scalar] =
+        val n = BigInt(BigInteger(s))
+        Scalar.apply(n)
+
+    // Conversions
+    def fromByteStringBigEndian(bytes: ByteString): Option[Scalar] =
+        Scalar.apply(byteStringToInteger(true, bytes))
+
+    def fromByteStringLittleEndian(bytes: ByteString): Option[Scalar] =
+        Scalar.apply(byteStringToInteger(false, bytes))
+
+    extension (self: Scalar)
+
+        // Combining
+
+        // Adds two `Scalar` elements, ensuring the result stays within the finite field range.
+        def add(addend: Scalar): Scalar =
+            new Scalar((self.unScalar + addend.unScalar) % fieldPrime)
+
```

Comment:
You can use `def +(add: Scala): Scala = Scalar((self.unScalar + addend.unScalar) % fieldPrime)